File: projectforge-caldav/src/main/java/org/projectforge/caldav/controller/ProjectforgeCarddavController.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.projectforge.caldav.controller;

import java.util.List;

import org.projectforge.caldav.config.ApplicationContextProvider;
import org.projectforge.caldav.model.AddressBook;
import org.projectforge.caldav.model.Contact;
import org.projectforge.caldav.model.ContactsHome;
import org.projectforge.caldav.model.User;
import org.projectforge.caldav.model.UsersHome;
import org.projectforge.caldav.rest.AddressRest;

import io.milton.annotations.AddressBooks;
import io.milton.annotations.ChildrenOf;
import io.milton.annotations.ContactData;
import io.milton.annotations.Delete;
import io.milton.annotations.Get;
import io.milton.annotations.PutChild;
import io.milton.annotations.ResourceController;
import io.milton.annotations.Root;

@ResourceController
public class ProjectforgeCarddavController extends BaseDavController
{
  private static org.slf4j.Logger log = org.slf4j.LoggerFactory.getLogger(ProjectforgeCarddavController.class);

  private AddressRest addressRest;

  public ProjectforgeCarddavController()
  {
    log.info("ProjectforgeCarddavController()");
  }

  @Root
  public ProjectforgeCarddavController getRoot()
  {
    return this;
  }

  @ChildrenOf
  public UsersHome getUsersHome(ProjectforgeCarddavController root)
  {
    if (usersHome == null) {
      log.info("Create new UsersHome");
      usersHome = new UsersHome();
    }
    return usersHome;
  }

  @ChildrenOf
  public ContactsHome getContactsHome(User user)
  {
    ContactsHome contactsHome = getUserCache().getUserContactsHomeMap().get(user.getPk());
    if (contactsHome != null) {
      return contactsHome;
    }
    log.info("Creating ContactsHome for user:" + user);
    contactsHome = new ContactsHome(user);
    getUserCache().getUserContactsHomeMap().put(user.getPk(), contactsHome);
    return contactsHome;
  }

  @ChildrenOf
  @AddressBooks
  public AddressBook getAddressBook(ContactsHome cons)
  {
    AddressBook addressBook = getUserCache().getUserAddressBookMap().get(cons.getUser().getPk());
    if (addressBook != null) {
      return addressBook;
    }
    log.info("Creating AddressBook for user:" + cons.getUser());
    addressBook = new AddressBook(cons.getUser());
    getUserCache().getUserAddressBookMap().put(cons.getUser().getPk(), addressBook);
    return addressBook;
  }

  @ChildrenOf
  public List<Contact> getContact(AddressBook ab)
  {
    return getAddressRest().getContactList(ab);
  }

  @Get
  @ContactData
  public byte[] getContactData(Contact c)
  {
    return c.getVcardData();
  }

  @PutChild
  public Contact createContact(AddressBook ab, byte[] vcardBytearray, String newName)
  {
    log.info("CreateContact: " + newName);
    Contact c = getAddressRest().createContact(ab, vcardBytearray);
    return c;
  }

  @PutChild
  public Contact updateContact(Contact c, byte[] vcardBytearray)
  {
    log.info("updateContact: " + c.getName());
    c = getAddressRest().updateContact(c, vcardBytearray);
    return c;
  }

  @Delete
  public void deleteContact(Contact c)
  {
    log.info("deleteContact: " + c.getName());
    getAddressRest().deleteContact(c);
  }

  private AddressRest getAddressRest()
  {
    if (addressRest == null) {
      addressRest = ApplicationContextProvider.getApplicationContext().getBean(AddressRest.class);
    }
    return addressRest;
  }

}

File: projectforge-caldav/src/main/kotlin/org/projectforge/caldav/controller/ProjectForgeCardDAVController.kt
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2022 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.caldav.controller

import io.milton.annotations.*
import mu.KotlinLogging
import org.projectforge.caldav.model.*
import org.projectforge.caldav.service.AddressService
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired

// curl -A "CoreDAV" --user "kai:xxxx" -H "Content-Type: text/xml" -H "Depth: 1" --data "<propfind xmlns='DAV:'><prop><address-data xmlns='urn:ietf:params:xml:ns:carddav'/></prop></propfind>" -X PROPFIND "http://localhost:8080/users/kai/addressBooks/default"

private val log = KotlinLogging.logger {}

@ResourceController
open class ProjectForgeCardDAVController : BaseDAVController() {
    @Autowired
    private lateinit var addressService_: AddressService

    private val addressService: AddressService
        get() {
            ensureAutowire()
            return addressService_
        }

    @get:Root
    val root: ProjectForgeCardDAVController
        get() = this

    @ChildrenOf
    fun getUsersHome(root: ProjectForgeCardDAVController?): UsersHome {
        if (usersHome == null) {
            log.info("Create new UsersHome")
            usersHome = UsersHome()
        }
        return usersHome!!
    }

    @ChildrenOf
    fun getContactsHome(user: User): ContactsHome {
        log.info("Creating ContactsHome for user:$user")
        return ContactsHome(user)
    }

    @ChildrenOf
    @AddressBooks
    fun getAddressBook(cons: ContactsHome): AddressBook {
        log.info("getAddressBook: '${cons.name}' for user '${cons.user.username}'.")
        return AddressBook(cons.user)
    }

    @ChildrenOf
    fun getContacts(ab: AddressBook): List<Contact> {
        log.info("getContacts for address book '${ab.name}' and user '${ab.user.username}'.")
        return addressService.getContactList(ab)
    }

    @Get
    @ContactData
    fun getContactData(c: Contact): ByteArray? {
        if (log.isDebugEnabled) {
            log.debug("getContactData: '${c.name}' with id ${c.id}.")
        }
        return c.vcardData
    }

    @PutChild
    fun createContact(ab: AddressBook, vcardBytearray: ByteArray, newName: String): Contact {
        log.info("CreateContact: $newName")
        return addressService.createContact(ab, vcardBytearray)
    }

    @PutChild
    fun updateContact(c: Contact, vcardBytearray: ByteArray?): Contact {
        log.info("updateContact: '${c.name}' with id ${c.id}.")
        return addressService.updateContact(c, vcardBytearray)
    }

    @Delete
    fun deleteContact(c: Contact) {
        log.info("deleteContact: '${c.name}' with id ${c.id}.")
        addressService.deleteContact(c)
    }

    init {
        log.info("init")
    }
}
